### ISSUE TITLE
Support user lock-out after X number of failed sign-in attempts in Y time

### ISSUE THREAD

**Issue by vsimonian** [anvilresearch/connect#112]
I think it would be worthwhile to implement a configurable feature that would allow a user to be locked out if there are X number of failed sign-in attempts in Y time. The lock-out would expire after Z amount of time, or when the administrator unlocked the user manually, whichever comes first. Lastly, the lock-out could be configured not to expire on its own and require manual intervention.

I believe a feature like this would put Connect more within the realm of requirements that a corporate environment may have.

To be clear, this is a proposition for the password provider (local strategy), not external providers.

**Comment by christiansmith** [comment]
Agreed. Related issues:

#18 Endpoint rate limiting
#20 Password reset
#21 Password signup verification (email verification)
#47 Sign in via third party where there is an existing account with the same email
#48 Alternate email for signin
#49 Recovery email

Could be a good set of stuff to bang out in a weekend hackathon.

**Comment by bnb** [comment]
One implementation of this that I've seen is to escalate the blacklist period as more and more attempts are failed. 6 fails = 1 minute, 10 fails = 5 minutes, 16 fails = 1 hour, 20 fails = admin-approval required. Also emphasizing a Forgot Password link would be effective to help actual users facing this problem.

**Comment by bnb** [comment]
However, thinking about that again, it would be simple to lock anyone out of their account. Some system for this would need to be put in place, and is probably a lot more complex than we can do.

**Comment by vsimonian** [comment]
I agree that, given the right configuration, it would be possible to intentionally lock an arbitrary user out. However, this is a fundamental limitation of a lock-out system in general more so than it is of implementing it in Connect. Nevertheless, certain enterprises have strict security policies in place, and having this as an optional feature would allow Connect to be deployed in more environments.

Could you elaborate a bit on "more complex than we can do," @bnb? I'm not quite following you completely on that point.

**Comment by bnb** [comment]
@vsimonian By that I just meant the logic and code behind restricting access to the right place at the right time - i.e. limiting users with certain traits (IP, IP range, browser config, connection type, and so on) but not others, which is the way I imagined this would be solved.

**Comment by vsimonian** [comment]
@bnb Ah, I think I follow you now. I believe in order to avoid any unnecessary complexities, the lock-out function should be made simple and affect a `lastFailedSignin` or other property of some sort on the user object no matter what the connection. Typically, a lock-out mechanism is only deployed in a controlled environment, or an environment where even a lock-out is preferable compared to a successful sign-in. Paired with an effective rate limiter, I doubt it will be easy for an attacker to do sufficient damage to the company in question.

**Comment by vsimonian** [comment]
An afterthought - I think it would also be nice to have the option of having an e-mail sent to the user with a tokenized link to a page where they can revoke the lock-out themselves.

**Comment by Cynfusion** [comment]
+1 @christiansmith on your list, those are going to be necessary early on
